File: exchange/fetcher.go
package exchange

import (
	"context"
	"sync"
	"time"

	"crypto-tray/config"
	"crypto-tray/internal/httpclient"
)

// Fetcher periodically fetches exchange rates
type Fetcher struct {
	configManager  *config.Manager
	callback       RatesHandler
	primaryClient  *httpclient.Client
	fallbackClient *httpclient.Client

	mu           sync.RWMutex
	cancelFunc   context.CancelFunc
	refreshCh    chan struct{}
	currentRates *ExchangeRates
}

// newFetcher creates a new exchange rate fetcher (internal use by Service)
func newFetcher(configManager *config.Manager) *Fetcher {
	return &Fetcher{
		configManager: configManager,
		primaryClient: httpclient.New(httpclient.Config{
			BaseURL: primaryURL,
			Timeout: timeout,
		}),
		fallbackClient: httpclient.New(httpclient.Config{
			BaseURL: fallbackURL,
			Timeout: timeout,
		}),
		refreshCh: make(chan struct{}, 1),
	}
}

// Start begins the exchange rate fetching loop with the given callback
func (f *Fetcher) Start(callback RatesHandler) {
	f.mu.Lock()
	f.callback = callback
	f.mu.Unlock()

	ctx, cancel := context.WithCancel(context.Background())

	f.mu.Lock()
	f.cancelFunc = cancel
	f.mu.Unlock()

	go f.loop(ctx)
}

// Stop stops the exchange rate fetching loop
func (f *Fetcher) Stop() {
	f.mu.Lock()
	if f.cancelFunc != nil {
		f.cancelFunc()
	}
	f.mu.Unlock()
}

// RefreshNow triggers an immediate exchange rate fetch
func (f *Fetcher) RefreshNow() {
	select {
	case f.refreshCh <- struct{}{}:
	default:
		// Already a refresh pending
	}
}

// GetRates returns the current cached exchange rates (thread-safe)
func (f *Fetcher) GetRates() *ExchangeRates {
	f.mu.RLock()
	defer f.mu.RUnlock()
	return f.currentRates
}

func (f *Fetcher) loop(ctx context.Context) {
	// Fetch immediately on start
	f.fetchOnce(ctx)

	cfg := f.configManager.Get()
	currentInterval := cfg.RefreshSeconds
	ticker := time.NewTicker(time.Duration(currentInterval) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			cfg = f.configManager.Get()
			f.fetchOnce(ctx)
			// Update ticker if interval changed
			if cfg.RefreshSeconds != currentInterval {
				currentInterval = cfg.RefreshSeconds
				ticker.Reset(time.Duration(currentInterval) * time.Second)
			}
		case <-f.refreshCh:
			cfg = f.configManager.Get()
			f.fetchOnce(ctx)
			// Reset ticker after manual refresh
			currentInterval = cfg.RefreshSeconds
			ticker.Reset(time.Duration(currentInterval) * time.Second)
		}
	}
}

func (f *Fetcher) fetchOnce(ctx context.Context) {
	rates, err := f.fetchFromClient(ctx, f.primaryClient)
	if err != nil {
		rates, err = f.fetchFromClient(ctx, f.fallbackClient)
		if err != nil {
			// Keep using cached rates
			f.callback(f.GetRates(), nil)
			return
		}
	}

	// Update cache with successful fetch
	f.mu.Lock()
	f.currentRates = rates
	f.mu.Unlock()

	f.callback(rates, nil)
}

func (f *Fetcher) fetchFromClient(ctx context.Context, client *httpclient.Client) (*ExchangeRates, error) {
	var resp APIResponse

	if err := client.Get(ctx, "/"+baseCurrency+".json", &resp); err != nil {
		return nil, err
	}

	return &ExchangeRates{
		Date:  resp.Date,
		Base:  baseCurrency,
		Rates: resp.USDT,
	}, nil
}

File: tauri/src-tauri/src/exchange/fetcher.rs
use std::sync::Arc;
use std::time::Duration;
use tauri::{AppHandle, Emitter, Wry};
use tokio::sync::{mpsc, watch, RwLock};
use tokio::time::interval;

use crate::config::ConfigManager;

use super::types::{ApiResponse, ExchangeRates, BASE_CURRENCY, FALLBACK_URL, PRIMARY_URL, TIMEOUT};

pub struct Fetcher {
    current_rates: Arc<RwLock<Option<ExchangeRates>>>,
    refresh_tx: mpsc::Sender<()>,
}

impl Fetcher {
    pub fn new(
        app: AppHandle<Wry>,
        config_manager: Arc<ConfigManager>,
        shutdown_rx: watch::Receiver<bool>,
    ) -> Self {
        let (refresh_tx, refresh_rx) = mpsc::channel::<()>(1);
        let current_rates = Arc::new(RwLock::new(None));

        let client = reqwest::Client::builder()
            .timeout(TIMEOUT)
            .build()
            .unwrap_or_default();

        let fetcher = Self {
            current_rates: current_rates.clone(),
            refresh_tx,
        };

        tauri::async_runtime::spawn(Self::run_loop(
            app,
            config_manager,
            current_rates,
            refresh_rx,
            client,
            shutdown_rx,
        ));

        fetcher
    }

    pub async fn refresh_now(&self) {
        if let Err(e) = self.refresh_tx.send(()).await {
            eprintln!("Failed to send exchange refresh signal: {}", e);
        }
    }

    pub async fn get_rates(&self) -> Option<ExchangeRates> {
        self.current_rates.read().await.clone()
    }

    async fn run_loop(
        app: AppHandle<Wry>,
        config_manager: Arc<ConfigManager>,
        current_rates: Arc<RwLock<Option<ExchangeRates>>>,
        mut refresh_rx: mpsc::Receiver<()>,
        client: reqwest::Client,
        mut shutdown_rx: watch::Receiver<bool>,
    ) {
        Self::fetch_once(&current_rates, &client).await;
        Self::emit_rates(&app, &current_rates).await;

        let cfg = config_manager.get();
        let mut current_interval = cfg.refresh_seconds;
        let mut poll_interval = interval(Duration::from_secs(current_interval as u64));

        loop {
            if *shutdown_rx.borrow() {
                return;
            }

            tokio::select! {
                _ = poll_interval.tick() => {
                    let cfg = config_manager.get();
                    Self::fetch_once(&current_rates, &client).await;
                    Self::emit_rates(&app, &current_rates).await;

                    if cfg.refresh_seconds != current_interval {
                        current_interval = cfg.refresh_seconds;
                        poll_interval = interval(Duration::from_secs(current_interval as u64));
                    }
                }
                Some(()) = refresh_rx.recv() => {
                    let cfg = config_manager.get();
                    Self::fetch_once(&current_rates, &client).await;
                    Self::emit_rates(&app, &current_rates).await;

                    current_interval = cfg.refresh_seconds;
                    poll_interval = interval(Duration::from_secs(current_interval as u64));
                }
                result = shutdown_rx.changed() => {
                    if result.is_err() || *shutdown_rx.borrow() {
                        return;
                    }
                }
            }
        }
    }

    async fn fetch_once(
        current_rates: &Arc<RwLock<Option<ExchangeRates>>>,
        client: &reqwest::Client,
    ) {
        let url = format!("{}/{}.json", PRIMARY_URL, BASE_CURRENCY);
        match Self::fetch_from_url(client, &url).await {
            Ok(rates) => {
                *current_rates.write().await = Some(rates);
                return;
            }
            Err(_) => {
                let fallback_url = format!("{}/{}.json", FALLBACK_URL, BASE_CURRENCY);
                if let Ok(rates) = Self::fetch_from_url(client, &fallback_url).await {
                    *current_rates.write().await = Some(rates);
                }
            }
        }
    }

    async fn emit_rates(app: &AppHandle<Wry>, current_rates: &Arc<RwLock<Option<ExchangeRates>>>) {
        let rates = current_rates.read().await;
        if let Some(ref rates) = *rates {
            if let Err(e) = app.emit("exchange:update", &rates.rates) {
                eprintln!("Failed to emit exchange update: {}", e);
            }
        }
    }

    async fn fetch_from_url(
        client: &reqwest::Client,
        url: &str,
    ) -> Result<ExchangeRates, reqwest::Error> {
        let resp: ApiResponse = client.get(url).send().await?.json().await?;

        Ok(ExchangeRates {
            date: resp.date,
            base: BASE_CURRENCY.to_string(),
            rates: resp.usdt,
        })
    }
}
